### ISSUE TITLE
[Help Wanted] [PLT-6074] Have deleteReaction API send the correct value for post.HasReactions

### ISSUE THREAD

**Issue by jasonblais** [mattermost/platform#6278]
If you're interested please comment here and come [join our "Contributors" community channel](https://pre-release.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://pre-release.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://docs.mattermost.com/guides/developer.html), specifically for [machine setup](https://docs.mattermost.com/developer/developer-setup.html) and for [developer workflow](https://docs.mattermost.com/developer/developer-flow.html).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/PLT-5576)

The changes made in https://github.com/mattermost/platform/pull/5499 make it so that the websocket event fired after removing a reaction always have the client save the post.HasReactions field to true even when the last reaction has just been removed. Since that fix was made to solve a read replica issue (caused by saving the updated post on master and then getting the post again from a slave), we should change it so that the SqlReactionStore.Delete method returns the updated post instead of the the reaction that was just deleted (since the caller should already have that). The SqlReactionStore.Save method should probably be updated to return the post as well to keep the behaviour parallel between the two.

This is also causing https://mattermost.atlassian.net/browse/PLT-5941, but the community member working on it came up with an alternate fix for the web app that we'll be able to get in for 3.8

**Comment by difarem** [comment]
I'd like to have a go at this. Just to be absolutely clear: the objective is to have the internal reaction adding and removing methods return the updated Post instead of the Reaction, correct?

**Comment by difarem** [comment]
WIP: https://github.com/difarem/mattermost-server/tree/PLT-6074

**Comment by hmhealey** [comment]
@difarem To clarify, whenever a reaction is saved or deleted right now, it sends the client two websocket events ([link](https://github.com/mattermost/mattermost-server/blob/master/app/reaction.go#L50)): a reaction_added/reaction_removed event depending on what happened and a post_edited event to tell the client that the post may now have reactions.

The problem with with the post_edited event. Regardless of how many reactions the post has (like if the only reaction were removed), it sends the post to the client with post.HasReactions set to true ([link](https://github.com/mattermost/mattermost-server/blob/master/app/reaction.go#L62)).

 The fix that I proposed in the ticket is close to what you suggested, but it would be to figure out how to return the post from the database when the reaction is deleted in `app.DeleteReactionForPost`, so that we can pass it into `app.sendReactionEvent` to send it off to the client. It's been a while since I've worked on the store code, but I think the function responsible for deleting the reaction [here](https://github.com/mattermost/mattermost-server/blob/master/store/sqlstore/supplier_reactions.go#L62).

One thing to be careful of is that we read/write off of different databases for performance reasons, so saving the post and then reading it back immediately may not work. This code used to do that and had that exact issue.

**Comment by jasonblais** [comment]
@difarem Is there anything we can help with?

**Comment by difarem** [comment]
@jasonblais not really, I don't think I'll finish this.

**Comment by jasonblais** [comment]
No worries, thanks @difarem - if you're interested working on it later, or [on another Go ticket](https://github.com/mattermost/mattermost-server/labels/Go), let us know and we'd be happy to help with any questions.

**Comment by pradeepmurugesan** [comment]
@jasonblais Is this ticket still valid.. Shall I start working on it ?

**Comment by jasonblais** [comment]
Yes it's still valid, would love to have your help with it @pradeepmurugesan!

**Comment by pradeepmurugesan** [comment]
@jasonblais Looking at the comments and task description I got the problem as follows.

The post edited web socket event has the field ```hasReactions``` set to true even if its the last reaction that is getting removed ? So we need to send ```hasReactions``` as false when the last reaction is removed.

Kindly let me know If I am getting the task here. 

Pradeep.

**Comment by hmhealey** [comment]
@pradeepmurugesan That's correct. We used to have that functionality, but there was a bug that happened in some database setups, so we had to remove it. There's some more details in the Jira ticket linked above that go into some more detail about it.

**Comment by pradeepmurugesan** [comment]
@hmhealey Cool.. I understand that reading the DB after writing will take some time due to the 2 different Dbs for read and write... 

I am trying to check if the post has the only one reaction at the time it comes to ```DeleteReactionForPost```. If so set the flag ```hasReactions``` to false else we leave it to default (true).
Do you think this will work or see any other side effects?  

Sorry I am asking too many questions. I am new to both goLang and the project.

**Comment by hmhealey** [comment]
No worries about the questions. I'm glad to help.

I think that should work. There's a small chance that you could have two people delete their reactions at nearly the same time so that the websocket passes back `HasReactions = true` for both of them, but that seems extremely unlikely.

The only other thing I could think to do would be to have the SQL statement in the store return whether or not it changed HasReactions on the post when it goes to make the change, but that would be more complicated to solve the rare case I mentioned above.

**Comment by pradeepmurugesan** [comment]
@hmhealey True.. I thought about the complicated case, I too felt it will be hard to implement with the setup ( read and write from different DBs).

I have made the changes and created a PR. Kindly review.  

I couldn't find any Automated tests for this piece of code.. I have tested it manually and it works.

The only thing I found odd is if the hasReactions is set to false it never reaches the client. I assumed somewhere down the line the serializer removes the falsy flags. Kindly clarify.

**Comment by hmhealey** [comment]
It's left out by the serializer because Go uses something that it calls struct tags to define some things including how the json serializer should handle fields. If you look at `model/post.go` where the `HasReactions` field of the `Post` struct is defined ([link](https://github.com/mattermost/mattermost-server/blob/master/model/post.go#L81)), it has 
~~~
`json:"has_reactions,omitempty"`
~~~
after the type which means that the json encoder should store the name the field `has_reactions` and omit it if it's falsy. There's more info on that in the docs for the encoding/json package ([link](https://golang.org/pkg/encoding/json/#Marshal)).

**Comment by pradeepmurugesan** [comment]
Thanks..  That explains.. :-) 👍

**Comment by hanzei** [comment]
https://github.com/mattermost/mattermost-server/pull/9053 has been merged. Can this issue be closed?
